listener/src/serumListener.ts

```diff
@@ -0,0 +1,237 @@
+
+type ActiveSubscription = {
+  market: Market;
+  subscriptionId: number;
+}
+
+const addOpenOrdersIfMissing = async (connection: Connection, serumProgramId: PublicKey, openOrdersKey: PublicKey) => {
+  const { error, response } = await findOpenOrderByAddress(openOrdersKey.toString())
+  if (response) {
+    const { data } = await response.json()
+    if (!data.open_order_accounts.length) {
+      // get the open order account info from the chain
+      const openOrders = await OpenOrders.load(connection, openOrdersKey, serumProgramId);
+      upsertOpenOrder(openOrders)
+    }
+  }
+}
+
+function divideBnToNumber(numerator: BN, denominator: BN): number {
+  const quotient = numerator.div(denominator).toNumber();
+  const rem = numerator.umod(denominator);
+  const gcd = rem.gcd(denominator);
+  return quotient + rem.div(gcd).toNumber() / denominator.div(gcd).toNumber();
+}
+
+const _getFillSize = (market: Market, event: Event) => {
+  return divideBnToNumber(
+    event.eventFlags.bid ? event.nativeQuantityReleased : event.nativeQuantityPaid,
+    (market as any)._baseSplTokenMultiplier
+  )
+}
+
+const _getFillPrice = (market: Market, event: Event) => {
+  let priceBeforeFees
+
+  if (event.eventFlags.bid) {
+    priceBeforeFees = event.eventFlags.maker
+      ? event.nativeQuantityPaid.add(event.nativeFeeOrRebate)
+      : event.nativeQuantityPaid.sub(event.nativeFeeOrRebate)
+  } else {
+    priceBeforeFees = event.eventFlags.maker
+      ? event.nativeQuantityReleased.sub(event.nativeFeeOrRebate)
+      : event.nativeQuantityReleased.add(event.nativeFeeOrRebate)
+  }
+
+  const price = divideBnToNumber(
+    priceBeforeFees.mul((market as any)._baseSplTokenMultiplier),
+    (market as any)._quoteSplTokenMultiplier.mul(
+      event.eventFlags.bid ? event.nativeQuantityReleased : event.nativeQuantityPaid
+    )
+  )
+
+  return price
+}
+
+const _mapEventToDataMessage = (
+  event: Event,
+  market: Market,
+  timestamp: string,
+  slot: number,
+): Fill | Done | Change | undefined => {
+  const clientId = (event as any).clientOrderId ? (event as any).clientOrderId.toString() : undefined
+
+  const side = event.eventFlags.bid ? 'buy' : 'sell'
+  const orderId = event.orderId.toString()
+  const openOrdersAccount = event.openOrders.toBase58()
+  const openOrdersSlot = event.openOrdersSlot
+  const feeTier = event.feeTier
+
+  if (event.eventFlags.fill) {
+    const fillMessage: Fill = {
+      type: 'fill',
+      serumMarketAddress: market.address.toString(),
+      timestamp,
+      slot,
+      orderId,
+      clientId,
+      side,
+      price: _getFillPrice(market, event),
+      size: _getFillSize(market, event),
+      maker: event.eventFlags.maker,
+      feeCost: market.quoteSplSizeToNumber(event.nativeFeeOrRebate) * (event.eventFlags.maker ? -1 : 1),
+      account: openOrdersAccount,
+      accountSlot: openOrdersSlot,
+      feeTier: feeTier
+    }
+    return fillMessage
+  } else if (event.nativeQuantityPaid.eqn(0)) {
```

Comment:
nit: no need for the else since the other if returns

Reply:
It's a conditional, so it has purpose